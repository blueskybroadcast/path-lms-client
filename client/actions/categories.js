import normalize from 'jsonapi-normalizer';

import { FETCH_CATEGORIES, DELETE_CATEGORY } from './types';

export const fetchCategories = () => async (dispatch, getState, api) => {
  const res = await api.get('/categories');

  dispatch({
    type: FETCH_CATEGORIES,
    payload: normalize(res.data)
  });
};

export const addCategory = category => async (dispatch, getState, api) => {
  const { name, parentId, hidden } = category;

  await api.post('/categories', {
    name,
    parent_id: parentId,
    hidden
  }).then(() => dispatch(fetchCategories()));
};

export const deleteCategory = id => async (dispatch, getState, api) => {
  await api.delete(`/categories/${id}`).then(() => {
    dispatch({
      type: DELETE_CATEGORY,
      payload: id
    });
  });
};

export const editCategory = (id, category) => async (dispatch, getState, api) => {
  const { name, parentId, hidden } = category;

  await api.put(`/categories/${id}`, {
    name,
    parent_id: parentId,
    hidden
  }).then(() => {
    dispatch(fetchCategories());
  });
};
