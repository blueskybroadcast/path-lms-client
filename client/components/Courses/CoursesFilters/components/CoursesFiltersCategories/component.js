import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { Dropdown } from 'semantic-ui-react';

const CoursesFiltersCategories = ({
  categories, handleCategorySelect, handleCategoriesReset
}) => {
  const DropdownCategory = ({ category, children }) => (
    <li
      className={classNames({
        italic: category.hidden,
        'dropdown-submenu': category.children.length
      })}
    >
      <a onClick={() => handleCategorySelect(category.id)}>
        {category.name}
      </a>
      { children }
    </li>
  );

  return (
    <Dropdown
      text="Browse..."
      className="category-filter-trigger dropdown-trigger"
    >
      <Dropdown.Menu>
        <ul className="dropdown-menu">
          <li>
            <a onClick={handleCategoriesReset}>
              View all
            </a>
          </li>
          <li className="divider" />
          { categories.map(levelZero => (
            <DropdownCategory category={levelZero} key={levelZero.id}>
              { levelZero.children.length ? (
                <ul className="dropdown-menu">
                  { levelZero.children.map(levelOne => (
                    <DropdownCategory category={levelOne} key={levelOne.id}>
                      { levelOne.children.length ? (
                        <ul className="dropdown-menu">
                          { levelOne.children.map(levelTwo => (
                            <DropdownCategory category={levelTwo} key={levelTwo.id} />
                          ))}
                        </ul>
                      ) : null }
                    </DropdownCategory>
                  ))}
                </ul>
              ) : null}
            </DropdownCategory>
          ))}
        </ul>
      </Dropdown.Menu>
    </Dropdown>
  );
};

CoursesFiltersCategories.propTypes = {
  categories: PropTypes.arrayOf(PropTypes.object).isRequired,
  handleCategorySelect: PropTypes.func.isRequired,
  handleCategoriesReset: PropTypes.func.isRequired
};

export default CoursesFiltersCategories;
