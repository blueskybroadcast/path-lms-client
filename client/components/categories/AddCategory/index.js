import AddCategory from './container';

export default AddCategory;
