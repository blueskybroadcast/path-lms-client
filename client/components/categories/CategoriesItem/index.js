import CategoriesItem from './container';

export default CategoriesItem;
