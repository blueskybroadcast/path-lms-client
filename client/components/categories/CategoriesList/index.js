import CategoriesList from './container';

export default CategoriesList;
