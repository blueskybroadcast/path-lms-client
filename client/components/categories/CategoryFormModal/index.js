import CategoryFormModal from './container';

export default CategoryFormModal;
