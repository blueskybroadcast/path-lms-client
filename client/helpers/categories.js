export const getParentListLevelOne = (categories) => {
  const parentsLevelOne = categories.filter(cat => cat.level === 0 && cat.childCount > 0);
  return parentsLevelOne.map(({ name, id }) => ({
    name,
    list: categories.filter(cat => cat.parentId === id)
  }));
};

export const getSortedCategories = (categoriesIds, categoriesData) => {
  const categories = [];

  categoriesIds.forEach((id) => {
    const data = categoriesData[id];
    if (!data.parent.id) {
      categories.push({
        id,
        level: 0,
        name: data.name,
        childCount: 0,
        hidden: data.hidden
      });
    }
  });

  const nextCategories = [...categories];
  let nextAddedCategoriesCount = 0;

  categories.forEach((cat, index) => {
    categoriesIds.forEach((id) => {
      const data = categoriesData[id];
      const parentId = data.parent.id;

      if (cat.id === parentId) {
        nextCategories.splice(
          index + 1 + nextAddedCategoriesCount,
          0,
          {
            id,
            level: 1,
            name: data.name,
            childCount: 0,
            parentId,
            hidden: data.hidden
          }
        );
        categories[index].childCount += 1;
        nextAddedCategoriesCount += 1;
      }
    });
  });

  const finalCategories = [...nextCategories];
  let finalAddedCategoriesCount = 0;

  nextCategories.forEach((cat, index) => {
    if (cat.level === 1) {
      categoriesIds.forEach((id) => {
        const data = categoriesData[id];
        if (cat.id === data.parent.id) {
          finalCategories.splice(
            index + 1 + finalAddedCategoriesCount,
            0,
            {
              id,
              level: 2,
              name: data.name,
              hidden: data.hidden
            }
          );
          nextCategories[index].childCount += 1;
          finalAddedCategoriesCount += 1;
        }
      });
    }
  });

  return finalCategories;
};

export const getCategoriesForSelect = (categoriesIds, categoriesData) => {
  if (!categoriesIds.length) return [];
  const remainingCategories = [...categoriesIds];
  const categories = [];

  categoriesIds.forEach((id) => {
    const data = categoriesData[id];
    if (!data.parent.id) {
      categories.push({
        id,
        name: data.name,
        hidden: data.hidden,
        children: []
      });
      const itemToRemoveIndex = remainingCategories.indexOf(id);
      remainingCategories.splice(itemToRemoveIndex, 1);
    }
  });

  const finalCategories = [...categories];

  categories.forEach((cat, index) => {
    remainingCategories.forEach((id) => {
      const data = categoriesData[id];
      const parentId = data.parent.id;

      if (cat.id === parentId) {
        const levelOneCat = {
          id,
          name: data.name,
          hidden: data.hidden,
          children: []
        };

        remainingCategories.forEach((childId) => {
          const childData = categoriesData[childId];
          const childParentId = childData.parent.id;

          if (id === childParentId) {
            const levelTwoCat = {
              id: childId,
              name: childData.name,
              hidden: childData.hidden,
              children: []
            };

            levelOneCat.children.push(levelTwoCat);
          }
        });

        finalCategories[index].children.push(levelOneCat);
      }
    });
  });

  return finalCategories;
};
