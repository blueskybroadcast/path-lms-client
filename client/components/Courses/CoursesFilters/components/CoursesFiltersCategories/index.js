import CoursesFiltersCategories from './component';

export default CoursesFiltersCategories;
