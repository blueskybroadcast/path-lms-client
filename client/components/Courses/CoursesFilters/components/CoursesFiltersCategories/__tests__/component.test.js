import React from 'react';
import { shallow } from 'enzyme';
import CoursesFiltersCategories from '../component';

import { categoriesForSelect } from '../../../../../../fixtures/categories';

describe('Courses filters categories', () => {
  test('should render CoursesFiltersCategories with data correctly', () => {
    const wrapper = shallow(
      <CoursesFiltersCategories
        categories={categoriesForSelect}
        handleCategorySelect={jest.fn()}
        handleCategoriesReset={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render CoursesFiltersCategories without data correctly', () => {
    const wrapper = shallow(
      <CoursesFiltersCategories
        handleCategorySelect={jest.fn()}
        handleCategoriesReset={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
