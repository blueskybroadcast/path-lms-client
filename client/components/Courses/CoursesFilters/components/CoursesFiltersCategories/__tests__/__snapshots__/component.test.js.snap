// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Courses filters categories should render CoursesFiltersCategories with data correctly 1`] = `
<Dropdown
  additionLabel="Add "
  additionPosition="top"
  className="category-filter-trigger dropdown-trigger"
  closeOnBlur={true}
  deburr={false}
  icon="dropdown"
  minCharacters={1}
  noResultsMessage="No results found."
  openOnFocus={true}
  renderLabel={[Function]}
  searchInput="text"
  selectOnBlur={true}
  selectOnNavigation={true}
  text="Browse..."
  wrapSelection={true}
>
  <DropdownMenu>
    <ul
      className="dropdown-menu"
    >
      <li>
        <a
          onClick={[MockFunction]}
        >
          View all
        </a>
      </li>
      <li
        className="divider"
      />
      <DropdownCategory
        category={
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "hidden": false,
                "id": "37",
                "name": "Cat Top 1 child",
              },
            ],
            "hidden": true,
            "id": "36",
            "name": "Cat Top 1",
          }
        }
        key="36"
      >
        <ul
          className="dropdown-menu"
        >
          <DropdownCategory
            category={
              Object {
                "children": Array [],
                "hidden": false,
                "id": "37",
                "name": "Cat Top 1 child",
              }
            }
            key="37"
          />
        </ul>
      </DropdownCategory>
      <DropdownCategory
        category={
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "hidden": false,
                    "id": "34",
                    "name": "Cat Top 2 child child",
                  },
                ],
                "hidden": false,
                "id": "33",
                "name": "Cat Top 2 child",
              },
            ],
            "hidden": false,
            "id": "32",
            "name": "Cat Top 2",
          }
        }
        key="32"
      >
        <ul
          className="dropdown-menu"
        >
          <DropdownCategory
            category={
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "hidden": false,
                    "id": "34",
                    "name": "Cat Top 2 child child",
                  },
                ],
                "hidden": false,
                "id": "33",
                "name": "Cat Top 2 child",
              }
            }
            key="33"
          >
            <ul
              className="dropdown-menu"
            >
              <DropdownCategory
                category={
                  Object {
                    "children": Array [],
                    "hidden": false,
                    "id": "34",
                    "name": "Cat Top 2 child child",
                  }
                }
                key="34"
              />
            </ul>
          </DropdownCategory>
        </ul>
      </DropdownCategory>
    </ul>
  </DropdownMenu>
</Dropdown>
`;

exports[`Courses filters categories should render CoursesFiltersCategories without data correctly 1`] = `
<Dropdown
  additionLabel="Add "
  additionPosition="top"
  className="category-filter-trigger dropdown-trigger"
  closeOnBlur={true}
  deburr={false}
  icon="dropdown"
  minCharacters={1}
  noResultsMessage="No results found."
  openOnFocus={true}
  renderLabel={[Function]}
  searchInput="text"
  selectOnBlur={true}
  selectOnNavigation={true}
  text="Browse..."
  wrapSelection={true}
>
  <DropdownMenu>
    <ul
      className="dropdown-menu"
    >
      <li>
        <a
          onClick={[MockFunction]}
        >
          View all
        </a>
      </li>
    </ul>
  </DropdownMenu>
</Dropdown>
`;
